### PR TITLE
Feature:  OfType Operator

### PR DESCRIPTION
**Pull request by dwcullop**
## Description

A new operator `OfType` that is similar to the LINQ operator with the same name.  It converts a changeset of some base type (or interface) to a changeset of a different type by automatically excluding values that can't be cast as the new type.  It effectively combines `ImmutableFilter` and `ImmutableTransform` into a single operator for maximum efficiency for the case when both operations are trivial.

### Bonus Changes

Creates Test-compatible version of the `Spy` (named `TestSpy`) that emits output to `ITestOutputHelper` so that the contents of the Observables can be seen in the test results.

### CODE REVIEW

File: src/DynamicData.Tests/Utilities/ObservableSpy.cs [hunk 1]
```diff
@@ -125,6 +126,28 @@ private static class NativeMethods
     private static Func<IChangeSet<T>, string> CreateListChangeSetFormatter<T>(Func<T, string> formatter) where T : notnull =>
         cs => "[List Change Set]" + ChangeSetEntrySpacing + string.Join(ChangeSetEntrySpacing, cs.Select((change, n) => $"#{n} {FormatChange(formatter, change)}"));
 
+    public static IObservable<T> TestSpy<T>(this IObservable<T> source,
+                                                                ITestOutputHelper testOutputHelper, string? opName = null,
```
Comment (JakenVeina) [inline src/DynamicData.Tests/Utilities/ObservableSpy.cs:130]: This indentation makes my eyes twitch. :P

File: src/DynamicData/Cache/ObservableCacheEx.cs [hunk 2]
```diff
@@ -3316,6 +3316,26 @@ public static IObservable<T> FinallySafe<T>(this IObservable<T> source, Action f
         return source.Where(changes => changes.Count != 0);
     }
 
+    /// <summary>
+    /// Converts the changesets so that it only contains items of the given type.
+    /// </summary>
+    /// <typeparam name="TObject">The type of the object.</typeparam>
```
Comment (JakenVeina) [inline src/DynamicData/Cache/ObservableCacheEx.cs:3322]: Suggest "The type of items in the collection".

File: src/DynamicData/Cache/ObservableCacheEx.cs [hunk 3]
```diff
@@ -3316,6 +3316,26 @@ public static IObservable<T> FinallySafe<T>(this IObservable<T> source, Action f
         return source.Where(changes => changes.Count != 0);
     }
 
+    /// <summary>
+    /// Converts the changesets so that it only contains items of the given type.
```
Comment (JakenVeina) [inline src/DynamicData/Cache/ObservableCacheEx.cs:3320]: Suggest "Filters an observable collection to only include items of a specific type.", or something similar, as the description here.

File: src/DynamicData/Cache/ObservableCacheEx.cs [hunk 4]
```diff
@@ -3316,6 +3316,26 @@ public static IObservable<T> FinallySafe<T>(this IObservable<T> source, Action f
...
+    /// </summary>
+    /// <typeparam name="TObject">The type of the object.</typeparam>
+    /// <typeparam name="TKey">The type of the key.</typeparam>
+    /// <typeparam name="TDestination">The type of the objects for the result changeset.</typeparam>
```
Comment (JakenVeina) [inline src/DynamicData/Cache/ObservableCacheEx.cs:3324]: Suggest "The type of items to be allowed through the filter."

**Review by JakenVeina — APPROVED** [review]

**Review by RolandPheasant — APPROVED** [review]

File: src/DynamicData.Tests/Utilities/ObservableSpy.cs (hunk 1 above)
Comment (dwcullop) [reply src/DynamicData.Tests/Utilities/ObservableSpy.cs:130]: I'll flatten them out like all the others.

**Pull request merged**